
# 트랜잭션

## springframework - @Transactional 어노테이션

### 1.none-Transactional

![db-1](/etc/img/db-1.png)
- notice 테이블에는 4개의 row가 있습니다.
- Transactional 없이 notice_id 1을 가진 row를 삭제(delete)한뒤 notice_id 1을 가진 row를 조회한다면 어떻게 될까요?

![no-tran](/etc/img/no-tran.png)
- @Transactional을 없이 돌아가는 service logic입니다.
- jpa를 사용하였습니다
- deletbyId()는 notice_id를 통해 해당 notice를 지우겠다는 메서드입니다.
- findById()는 notice_id통해 해당 notice를 가져오겠따는 메서드입니다.
- 당연히 삭제후 조회는 되지 않아야합니다.

![exception-1](/etc/img/exception-1.png)
- 메서드를 실행시키면 exception이 발생하게 됩니다.
- 없는 Id를 조회하였습니다.
- 사진에는 없지만 없는 ID를 통해 조회하려고 하면 RunTimeException을 상속시킨 exception class를 발생시키게 만들었습니다.

![db-2](/etc/img/db-2.png)
- RunTimeException이 났지만 디비에는 row가 삭제 된걸 확인 할 수가 있습니다.
- 당연히 없는걸 조회하였으니 RunTimeException이 날수 밖에없다고 생각 할 수 도 있습니다.
- 하지만 이걸 추후 크게 확장해서 생각한다면 서버에서 RunTimeException이 발생했다는 것은 서버에 문제가 있다는 뜻입니다.
- 서버에 문제가 생겼는데 DB에 data를 정상적으로 처리 하는게 맞는걸까요?
- 좀 더 확장을 한다면 관계형 테이블(학생과 과목/다대다 관계 구성시)
- 학생테이블에 insert한후 과목테이블을 insert진행 중 RunTimeException이 발생되어 과목 테이블에 insert가 되지 않는다면??!! 
- 당연히 rollBack이 되어 처음상태로 돌려줘야합니다.
- 이러한 이유때문에 트랜잭션이 필요합니다.

### 2.Transactional

![db-1](/etc/img/db-1.png)
- notice 테이블에는 4개의 row가 있습니다.
- Transactional 있으면 notice_id 1을 가진 row를 삭제(delete)한뒤 notice_id 1을 가진 row를 조회한다면 어떻게 될까요?

![tran](/etc/img/tran.png)
- @Transactional을 있는 service logic입니다.
- jpa를 사용하였습니다

![exception-2](/etc/img/exception-2.png)
- 메서드를 실행시키면 exception이 발생하게 됩니다.

![db-3](/etc/img/db-3.png)
- RunTimeException이 났지만 디비에는 row가 삭제 된걸 확인 할 수가 있습니다.
- @Transactional 어노테이션을 이용하였기때문에 notice table row(id가 1인 row)가 삭제되지 않고 rollback되어있는걸 확인 할 수 있습니다.
- 트랜잭션은 runtimeExcetpion(unCheckedException) 발생시에만 rollback이 가능합니다.
- checkedException에도 rollback기능을 사용하고 싶으면 어노테이션에 추가적인 설정또는 
- checkedException을 runTimeException(unCheckedException)을 포장하여 발생시키면 가능합니다.
